### PR TITLE
Initial tests for Dijkstra's algorithm

### PR DESCRIPTION
**Pull request by zburk**
Assignment 6

#23  Initial tests for A6. I think they're all correct, but I'm human and there's a chance there is a mistake.

Changes proposed:
- Adding tests for A6

Tagging @benknoble

### CODE REVIEW

File: A6/DiGraphTests.java [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+
+    graph.addEdge(0, "a", "b", 1, null);
+
+    ShortestPathInfo[] results = graph.shortestPath("a");
+
+    assertEquals(0, results[0].getTotalWeight());
+    assertEquals("a", results[0].getDest());
+
+    assertEquals(1, results[1].getTotalWeight());
+    assertEquals("b", results[1].getDest());
```
Comment (benknoble) [inline A6/DiGraphTests.java]: What if mine come out in different orders?

**Comment by benknoble** [comment]
Note I didn't actually check your results to see if they are correct

File: A6/DiGraphTests.java (hunk 1 above)
Comment (zburk) [inline A6/DiGraphTests.java]: Confirmed on Piazza that they should be ordered by weight

**Comment by heyCarlH** [comment]
I checked it against my code and found test3 - test7 not pass. I passed all the tests in the oracle from last seamster. (change runTest import to f16 from s17.)

**Comment by zburk** [comment]
I am also passing all tests in the oracle from last semester

**Comment by heyCarlH** [comment]
OK. Then I should go to check my code.

File: A6/DiGraphTests.java (hunk 1 above)
Comment (zburk) [inline A6/DiGraphTests.java]: https://piazza.com/class/ixgyq2l7r9s8x?cid=259

**Comment by heyCarlH** [comment]
@Test
	public void algorithm3() {
		DiGraph graph = new DiGraph();
		graph.addNode(1, "a");
		graph.addNode(2, "b");
		graph.addNode(3, "c");
		graph.addNode(4, "d");

		graph.addEdge(0, "a", "b", 1, null);
		graph.addEdge(1, "b", "c", 1, null);
		graph.addEdge(2, "a", "d", 2, null);

		ShortestPathInfo[] results = graph.shortestPath("a");

		assertEquals(0, results[0].getTotalWeight());
		assertEquals("a", results[0].getDest());

		assertEquals(1, results[1].getTotalWeight());
		assertEquals("b", results[1].getDest());

		assertEquals(2, results[2].getTotalWeight());
		assertEquals("c", results[2].getDest());

		assertEquals(2, results[3].getTotalWeight());
		assertEquals("d", results[3].getDest());
	}

Here, the third item could be d or c; since they have the same total weight from a.

File: A6/DiGraphTests.java [hunk 2]
```diff
@@ -1,385 +1,392 @@
...
+		ShortestPathInfo[] results = graph.shortestPath("a");
+
+		assertEquals(0, results[0].getTotalWeight());
+		assertEquals("a", results[0].getDest());
+		
+		assertEquals(1, results[1].getTotalWeight());
+		assertEquals("b", results[1].getDest());
+		
+		assertEquals(2, results[2].getTotalWeight());
+		assertEquals("c", results[2].getDest());
```
Comment (benknoble) [inline A6/DiGraphTests.java]: Still not sure about this whole 'require this order' problem... Could we just use an inefficient algorithm to test if one array contains all the elements of the other? $O(n^2)$, but we don't care for the tests about efficiency.

pseudo code:
```
sol[] <- elements of solution
results <- method call
boolean all = true
for each (el in sol)
    boolean found = false
    for each (res in result)
         compare for equality: if true, found = true
    all &= found
assertTrue(all)
```

**Comment by benknoble** [comment]
@heyCarlH are they guaranteed to come out in *any* order?

**Comment by benknoble** [comment]
Aha! Just saw the Piazza post. Will re-review

File: A6/DiGraphTests.java (hunk 2 above)
Comment (benknoble) [inline A6/DiGraphTests.java]: Ignore this

**Comment by heyCarlH** [comment]
@benknoble I don't think so, since these two paths have the same weight. But if you guys believe that it is OK to have different display orders for paths with same weight, then I can edit the test locally to fit my implementation.

**Comment by heyCarlH** [comment]
And leave a comment on GitHub for people to know that difference is OK.

**Comment by benknoble** [comment]
[This piazza post](https://piazza.com/class/ixgyq2l7r9s8x?cid=259) verifies the order should be by length of shortest path. It does *not* yet detail what to do for nodes with no path (length $-1$), so I won't check that part yet

File: A6/DiGraphTests.java [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+		
+		assertEquals(2, results[3].getTotalWeight());
+		assertEquals("d", results[3].getDest());
+    }
```
Comment (benknoble) [inline A6/DiGraphTests.java]: Ordering: same length?

File: A6/DiGraphTests.java [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+		
+		assertEquals(2, results[2].getTotalWeight());
+		assertEquals("b", results[2].getDest());
+		
+		assertEquals(3, results[3].getTotalWeight());
+		assertEquals("c", results[3].getDest());
+		
+		assertEquals(3, results[4].getTotalWeight());
+		assertEquals("e", results[4].getDest());
+
```
Comment (benknoble) [inline A6/DiGraphTests.java]: Ordering: same weight?

File: A6/DiGraphTests.java [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+		
+		assertEquals(-1, results[4].getTotalWeight());
+		assertEquals("e", results[4].getDest());
+		
+		assertEquals(-1, results[5].getTotalWeight());
+		assertEquals("f", results[5].getDest());
+		
+		assertEquals(-1, results[6].getTotalWeight());
+		assertEquals("g", results[6].getDest());
+
```
Comment (benknoble) [inline A6/DiGraphTests.java]: Seems like we should just fix this to make sure it works no matter what order those come back in. Since the last weights will all be $-1$, we just have to adjust the dest checks

File: A6/DiGraphTests.java [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+		assertEquals(graph.numOfNodes, results.length);
```
Comment (benknoble) [inline A6/DiGraphTests.java]: Same issue

**Comment by zburk** [comment]
Turns out order doesn't matter. Will need to update tests to reflect that

File: A6/DiGraphTests.java [hunk 7]
```diff
@@ -0,0 +1,392 @@
...
+		graph.delNode("b");
+
+		ShortestPathInfo[] results = graph.shortestPath("a");
+
+		assertEquals(0, results[0].getTotalWeight());
+		assertEquals("a", results[0].getDest());
+		
+		assertEquals(1, results[1].getTotalWeight());
+		assertEquals("d", results[1].getDest());
+
```
Comment (benknoble) [inline A6/DiGraphTests.java]: What about b? weight = 2

File: A6/DiGraphTests.java [hunk 8]
```diff
@@ -0,0 +1,392 @@
...
+		assertEquals("a", results[0].getDest());
+		
+		assertEquals(1, results[1].getTotalWeight());
+		assertEquals("d", results[1].getDest());
+		
+		assertEquals(3, results[2].getTotalWeight());
+		assertEquals("c", results[2].getDest());
+		
+		assertEquals(4, results[3].getTotalWeight());
+		assertEquals("e", results[3].getDest());
```
Comment (benknoble) [inline A6/DiGraphTests.java]: Can go a -(2)-> b -(1)-> e for length 3

File: A6/DiGraphTests.java [hunk 9]
```diff
@@ -0,0 +1,392 @@
...
+				d.addEdge((int) (i * Math.random() * Math.random() * 100000), Integer.toString(i), Integer.toString(i-1), 0, null);
+				d.addEdge((int) (i * Math.random() * Math.random() * 100000), Integer.toString(i), Integer.toString(i-2), 0, null);
+			}
+		}
+		
+		ShortestPathInfo[] results = d.shortestPath("3");
+
+		// Not quite sure how to test. Mainly to test algorithm is efficient
+		assertEquals(d.numOfNodes, results.length);
+    }
```
Comment (benknoble) [inline A6/DiGraphTests.java]: I'd reword the comment to simply say something like `//verify that the code doesn't explode with large numbers`

File: A6/DiGraphTests.java [hunk 10]
```diff
@@ -0,0 +1,392 @@
...
+		assertEquals("a", results[4].getDest());
+		
+		assertEquals(10, results[5].getTotalWeight());
+		assertEquals("f", results[5].getDest());
+    }
+
+    @Test
+    public void algorithmStressTest() {
+    	DiGraph d = new DiGraph();
+		for (int i = 0; i < 100; i++) {
```
Comment (benknoble) [inline A6/DiGraphTests.java]: can we make that `100` a static constant to easily manipulate it? Like `NUM_NODES`

File: A6/DiGraphTests.java (hunk 2 above)
Comment (benknoble) [inline A6/DiGraphTests.java]: Alternatively, we could use a HashMap for this...

File: A6/DiGraphTests.java (hunk 7 above)
Comment (benknoble) [inline A6/DiGraphTests.java]: @zburk tagging you for my most recent comments. Ran through and actually verified the graphs/path lengths this time. Also some stylistic things.

Finally, we do need to consider that the order they come out in is of no concern; therefore, we will need to actually check if each of the requisite solution elements is somewhere in the results array

Comment (zburk) [inline A6/DiGraphTests.java]: element b was removed. These integration tests are designed to test multiple functions coming together to ensure everything is stable

File: A6/DiGraphTests.java (hunk 8 above)
Comment (zburk) [inline A6/DiGraphTests.java]: See comment above

File: A6/DiGraphTests.java [hunk 11]
```diff
@@ -386,7 +307,16 @@ public void algorithmStressTest() {
 		
 		ShortestPathInfo[] results = d.shortestPath("3");
 
-		// Not quite sure how to test. Mainly to test algorithm is efficient
+		// Verify that the code doesn't explode with large numbers
 		assertEquals(d.numOfNodes, results.length);
     }
+    
+    public boolean checkShortestPath(HashMap<String,Integer> correct, DiGraph D, String label){
+    	ShortestPathInfo [] paths = D.shortestPath(label);
```
Comment (benknoble) [inline A6/DiGraphTests.java]: can we also check that `paths.length == D.numNodes() == correct.size()` ?
